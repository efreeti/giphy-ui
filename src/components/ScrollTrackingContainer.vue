<template>
  <div>
    <slot />

    <div v-if="trackScrollEnd">
      <div class="scroll-end-tracker" role="scroll-end-tracker" v-observe-visibility="handleVisibilityChange"></div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Emit, Vue } from 'vue-property-decorator';

@Component
export default class ScrollTrackingContainer extends Vue {
  @Prop() public trackScrollEnd!: boolean;

  @Emit() scrollEnd() {}

  handleVisibilityChange(isVisible: boolean) {
    if (isVisible) {
      this.scrollEnd();
    }
  }
}
</script>

<style>
.scroll-end-tracker {
  height: 20px;
}
</style>
