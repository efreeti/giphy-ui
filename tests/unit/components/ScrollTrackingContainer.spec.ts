import Vue from 'vue';
import { mount } from '@vue/test-utils';
import { intersectionObserver } from '@shopify/jest-dom-mocks';

import '../config';

import ScrollTrackingContainer from '@/components/ScrollTrackingContainer.vue';


describe('ScrollTrackingContainer', () => {
  beforeEach(() => {
    intersectionObserver.mock();
  });

  afterEach(() => {
    intersectionObserver.restore();
  });

  it('should render content correctly', () => {
    expect(mount(ScrollTrackingContainer, {
      slots: {
        default: '<div>some content</div>',
      },
    }))
        .toMatchSnapshot();
  });

  it('should render tracker correctly', () => {
    expect(mount(ScrollTrackingContainer, {
      propsData: {
        trackScrollEnd: true,
      },
    }))
        .toMatchSnapshot();
  });

  it('should emit "scroll-end" when tracked and visible', async () => {
    const wrapper = mount(ScrollTrackingContainer, {
      propsData: {
        trackScrollEnd: true,
      },
    });

    await Vue.nextTick();

		intersectionObserver.simulate({
			target: wrapper.find('[role=scroll-end-tracker]').element,
			isIntersecting: true,
		});

		expect(wrapper.emitted()['scroll-end'].length).toEqual(1);
  });

  it('should not emit "scroll-end" when not tracked and visible', async () => {
    const wrapper = mount(ScrollTrackingContainer, {
      propsData: {
        trackScrollEnd: false,
      },
    });

    await Vue.nextTick();

		intersectionObserver.simulate({
			target: wrapper.find('[role=scroll-end-tracker]').element,
			isIntersecting: true,
		});

		expect(wrapper.emitted()['scroll-end']).toBeUndefined();
  });
});
