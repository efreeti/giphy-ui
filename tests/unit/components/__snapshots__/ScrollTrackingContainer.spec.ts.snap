// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollTrackingContainer should render content correctly 1`] = `
<div>
  <div>some content</div>
  <!---->
</div>
`;

exports[`ScrollTrackingContainer should render tracker correctly 1`] = `
<div>
  <div>
    <div role="scroll-end-tracker" class="scroll-end-tracker"></div>
  </div>
</div>
`;
